From 6a11fbebd8459a987a2e5878b468e5f1f9e0dd75 Mon Sep 17 00:00:00 2001
From: Alexei Khlebnikov <alexei.khlebnikov@gmail.com>
Date: Thu, 22 Feb 2018 12:21:49 +0100
Subject: [PATCH 1/2] Fix for: "25.0.50; auto-revert-mode breaks git rebase"
 (Bug#21559)

* lisp/autorevert.el (auto-revert-handler):
Set "revert-buffer-in-progress-p" flag before calling
"vc-refresh-state".

* lisp/vc/vc-git.el (vc-git-command):
If "revert-buffer-in-progress-p" flag is set, prepend
"GIT_OPTIONAL_LOCKS=0" to "process-environment".  (Bug#21559)
---
 lisp/autorevert.el | 3 ++-
 lisp/vc/vc-git.el  | 9 ++++++++-
 2 files changed, 10 insertions(+), 2 deletions(-)

diff --git lisp/autorevert.el lisp/autorevert.el
index 7b8302695f..13cc25043d 100644
--- lisp/autorevert.el
+++ lisp/autorevert.el
@@ -719,7 +719,8 @@ This is an internal function used by Auto-Revert Mode."
     ;; `preserve-modes' avoids changing the (minor) modes.  But we do
     ;; want to reset the mode for VC, so we do it manually.
     (when (or revert auto-revert-check-vc-info)
-      (vc-refresh-state))))
+      (let ((revert-buffer-in-progress-p t))
+        (vc-refresh-state)))))
 
 (defun auto-revert-tail-handler (size)
   (let ((modified (buffer-modified-p))
diff --git lisp/vc/vc-git.el lisp/vc/vc-git.el
index efe853e5ee..0cea5334f4 100644
--- lisp/vc/vc-git.el
+++ lisp/vc/vc-git.el
@@ -1564,7 +1564,14 @@ The difference to vc-do-command is that this function always invokes
          (or coding-system-for-read vc-git-log-output-coding-system))
 	(coding-system-for-write
          (or coding-system-for-write vc-git-commits-coding-system))
-        (process-environment (cons "GIT_DIR" process-environment)))
+        (process-environment
+         (append
+          `("GIT_DIR"
+            ;; Avoid repository locking during background operations
+            ;; (bug#21559).
+            ,@(when revert-buffer-in-progress-p
+                '("GIT_OPTIONAL_LOCKS=0")))
+          process-environment)))
     (apply 'vc-do-command (or buffer "*vc*") okstatus vc-git-program
 	   ;; https://debbugs.gnu.org/16897
 	   (unless (and (not (cdr-safe file-or-list))
-- 
2.19.0

